src/aspire/operators/diag_matrix.py

```diff
@@ -0,0 +1,622 @@
+                )
+
+            # Then the mathematical `DA` for diag D and matrix A,
+            # is expressed in code as `D@A`.
+            #
+            # The operation `DA` is known as Left Scaling,
+            # and means to scale the row A_i by d_i in D.
+            # This can be accomplished by broadcasting an elemental multiply.
+
+            res = self._data[..., np.newaxis] * other
+        else:
+            raise RuntimeError(f"__matmul__ not implemented for {type(other)}.")
+
+        return res
+
+    def __matmul__(self, other):
+        """
+        Operator overload for matrix multiply of `DiagMatrix` instances.
+        """
+
+        return self.matmul(other)
+
+    def __rmatmul__(self, lhs):
+        """
+        Compute the right matrix multiplication with a `DiagMatrix` instance,
+        and a numpy array, `lhs` @ `self`.
+
+        :param other: The lhs Numpy instance.
+        :return: Returns numpy array representing `other @ self`.
+        """
+
+        # Note, we should only hit this method when mixing DiagMatrix with numpy.
+        #   This is because if both a and b are DiagMatrix,
+        #   then a@b would be handled first by a.__matmul__(b), never reaching here.
+        #   If a is BlkDiagMatrix, then it will handle the conversion of DiagMatrix b.
+
+        if not isinstance(lhs, np.ndarray):
+            raise RuntimeError(f"__rmatmul__ not implemented for {type(lhs)}.")
+
+        # For now, lets just interop with 2D `other` arrays,
+        # assumed to represent matrices.
+        if lhs.ndim != 2:
+            raise ValueError(
+                f"DiagMatrix.matmul of ndarray only supports 2D, received {lhs.shape}."
+            )
+
+        # Then the mathematical `AD` for diag D and matrix A,
+        # is expressed in code as `A@D`.
+        #
+        # The operation `AD` is known as Right Scaling,
+        # and means to scale the column A_j by d_j in D.
+        # This can be accomplished by broadcasting an elemental multiply.
+        # In this case A is the variable `lhs`
+
+        return lhs * self._data[np.newaxis]
+
+    def mul(self, other):
+        """
+        Compute the elementwise multiplication of a `DiagMatrix` instance and a
+        scalar or another `DiagMatrix`.
+
+        :param other: The rhs in the multiplication..
+        :return: A `self` * `other` as type of `other`.
+        """
+        if isinstance(other, DiagMatrix):
+            res = DiagMatrix(self._data * other._data)
+        elif isinstance(other, np.ndarray):
+            res = self * DiagMatrix(other)
+        elif isinstance(other, BlkDiagMatrix):
+            if self.stack_shape != ():
+                raise RuntimeError(
+                    f"Mixed `mul` only implemented for singletons at this time, received {self.stack_shape}."
+                )
+            res = self.as_blk_diag(other.partition) * other
```

Comment:
The end result here is diagonal, so having it return a `DiagMatrix` would be more efficient.

Reply:
You are right. I think I did that so something in cov2d was more natural.  It was so long ago at this point I don't remember.  This  (tending to up cast to BlkDiagMatrix) was the case somewhere else as well, but I changed the other case too, so might as well get this one. 

Changed to return DiagMatrix.  In the process I added a `diag` method to BlkDiagMatrix.

Reply:
Cool.